File: frontend/src/nodes2/nodes/hiresfix.ts
import { defineAsyncComponent } from "vue";
import type { ComfyNodeData } from "@/types/comfy";
import type { NodeModule } from "@/nodes2/nodeRegistry";
import { registerStyledNode } from "@/nodes2/nodeStyle";
import { addFilDomWidget, unmountAllFilWidgets } from "@/nodes2/domWidgetHost";
import { createSyncedNodeState, findFilWidget, sanitizeWidgetValue } from "@/nodes2/util";
import { applyFxComposables } from "@/nodes2/applyFxComposables";

const HiResFixVue = defineAsyncComponent(() => import("@/components/nodes/HiResFix.vue"));

const numericDefaults: Record<string, number> = {
  upscale_by: 1.25, denoise: 0.56, iterations: 1, strength: 1.0,
  seed: 0, hires_steps: 12,
};
const stringDefaults: Record<string, string> = {
  upscale_type: "latent", hires_ckpt_name: "(use same)",
  latent_upscaler: "nearest-exact", pixel_upscaler: "", control_net_name: "",
  preprocessor: "none",
};
const boolDefaults: Record<string, boolean> = {
  use_same_seed: true, use_controlnet: false,
};
const HIDE = [
  ...Object.keys(numericDefaults), ...Object.keys(stringDefaults), ...Object.keys(boolDefaults),
];

export const hiresfixNode: NodeModule = {
  id: "FiLHighResFix",
  register(nodeType: unknown, _nodeData: ComfyNodeData): void {
    registerStyledNode(nodeType, {
      // Height kept LOW on purpose — computeSize() (~430px real content)
      // always wins via Math.max in domWidgetHost.ts, so a buffer above it
      // here would just be dead space at the bottom. Width is the actual
      // reason this floor exists (computeSize()'s own width guess ignores it).
      //
      // 380, not 320: the own-seed row is a readout plus three pills, and the
      // Russian captions ("Случайно", "Прошлый", "Новый фикс.") need ~370px
      // before FilSeedRow starts truncating them. Nodes saved narrower are
      // widened to this floor on load.
      minSize: [380, 300],
      family: "sampling",
      description: "Latent/pixel upscale + re-sample settings, packed into a script for FiLKSampler.",
      badges: [{ text: "hires", color: "#bb9af7", text_color: "#0b0e14" }],
    });

    const proto = nodeType as {
      prototype: {
        onNodeCreated?: (...a: unknown[]) => unknown;
        onConfigure?: (...a: unknown[]) => unknown;
        onRemoved?: (...a: unknown[]) => unknown;
      };
    };
    const p = proto.prototype;

    const syncAll = (node: unknown, target: Record<string, unknown>) => {
      for (const name of Object.keys(numericDefaults)) {
        target[name] = sanitizeWidgetValue(findFilWidget(node, name), "number", numericDefaults[name]);
      }
      for (const name of Object.keys(stringDefaults)) {
        target[name] = sanitizeWidgetValue(findFilWidget(node, name), "string", stringDefaults[name]);
      }
      for (const name of Object.keys(boolDefaults)) {
        target[name] = sanitizeWidgetValue(findFilWidget(node, name), "boolean", boolDefaults[name]);
      }
    };

    const originalCreated = p.onNodeCreated;
    p.onNodeCreated = function (this: unknown, ...args: unknown[]) {
      const result = originalCreated?.apply(this, args);
      const node = this as { widgets?: unknown[]; _filHiResFixState?: unknown };
      const initial: Record<string, unknown> = {};
      syncAll(node, initial);
      for (const name of HIDE) {
        const w = findFilWidget(node, name);
        if (w) (w as { hidden?: boolean }).hidden = true;
      }
      // ComfyUI auto-adds a control_after_generate combo for the "seed"
      // widget. The panel owns seed UX (same as FiLSeed/FiLOpticScanner):
      // hide it and default it to "fixed" (matches the default same-seed
      // mode). HiResFix.vue then drives its value from the panel — "fixed"
      // in same-seed/own+fixed, "randomize" in own+random — so core itself
      // varies the hidden seed each queue (the mechanism that actually fires
      // on queue in this frontend version).
      const controlWidget = findFilWidget(node, "control_after_generate");
      if (controlWidget) {
        controlWidget.hidden = true;
        controlWidget.value = "fixed";
      }
      // `seed_mode` is panel-only state (no native widget behind it — the
      // Proxy mirror finds nothing to sync, which is fine); persists via
      // the DOM widget's getValue/setValue like scanner's.
      initial.seed_mode = "random";
      const state = {
        nodeState: createSyncedNodeState(node, initial),
        initialValues: { ...initial },
        ui: {},
        lastRunSeed: null,
      };
      Object.defineProperty(state, "node", { value: node, enumerable: false, configurable: true });
      node._filHiResFixState = state;
      addFilDomWidget(node, "fil_hiresfix_view", HiResFixVue, { state, height: 420 });
      return result;
    };

    const originalConfigure = p.onConfigure;
    p.onConfigure = function (this: unknown, ...args: unknown[]) {
      const result = originalConfigure?.apply(this, args);
      const node = this as { widgets?: unknown[]; _filHiResFixState?: { nodeState: Record<string, unknown> } };
      const state = node._filHiResFixState;
      if (!state) return result;
      syncAll(node, state.nodeState);
      return result;
    };

    const originalRemoved = p.onRemoved;
    p.onRemoved = function (this: unknown, ...args: unknown[]) {
      unmountAllFilWidgets(this);
      return originalRemoved?.apply(this, args);
    };

    applyFxComposables(nodeType as { prototype?: unknown });
  },
};

File: common/contracts/nodes/hiresfix.py
"""Frontend contract for `FiLHighResFix`."""

from __future__ import annotations

from ..widgets import _bool, _combo, _int, _segmented, _slider
from ..schema import NodeContract, NodeInputs, NodeOutput
from ...brand import CATEGORY_SAMPLING

CONTRACT = NodeContract(
    id="FiLHighResFix",
    title="🔬 HighRes Fix",
    category=CATEGORY_SAMPLING,
    description="Packs latent/pixel upscale + re-sample settings into a script.",
    # Keep in step with nodes2/nodes/hiresfix.ts — test_ui_presentation guards it.
    min_size=(300, 230),
    family="sampling",
    inputs=NodeInputs(
        required=[
            _segmented("upscale_type", options=["latent", "pixel", "both"], default="latent", label="Upscale type"),
            _combo("hires_ckpt_name", values=["(use same)"],
                   default="(use same)", label="Hires checkpoint"),
            _combo("latent_upscaler",
                   values=["nearest-exact", "bilinear", "area", "bicubic", "bislerp"],
                   default="nearest-exact", label="Latent upscaler",
                   visible_when="upscale_type", visible_when_value="latent"),
            _combo("pixel_upscaler", values=["(none)"],
                   default="(none)", label="Pixel upscaler",
                   visible_when="upscale_type", visible_when_value=["pixel", "both"]),
            _slider("upscale_by", default=1.25, minv=0.01, maxv=8.0, step=0.05, label="Upscale by"),
            _bool("use_same_seed", default=True, label="Use same seed"),
            _int("seed", default=0, minv=0, maxv=0xFFFFFFFFFFFFFFFF, step=1, label="Seed"),
            _int("hires_steps", default=12, minv=1, maxv=10000, step=1, label="Hires steps"),
            _slider("denoise", default=0.56, minv=0.0, maxv=1.0, step=0.01, label="Denoise"),
            _int("iterations", default=1, minv=0, maxv=5, step=1, label="Iterations"),
        ],
        optional=[
            _bool("use_controlnet", default=False, label="Use ControlNet", section="controlnet"),
            _combo("control_net_name", values=["(none)"],
                   default="(none)", label="ControlNet", section="controlnet",
                   visible_when="use_controlnet", visible_when_value=True),
            _slider("strength", default=1.0, minv=0.0, maxv=10.0, step=0.01, label="Strength",
                    section="controlnet", visible_when="use_controlnet", visible_when_value=True),
            _combo("preprocessor", values=["none", "canny"],
                   default="none", label="Preprocessor", section="controlnet",
                   visible_when="use_controlnet", visible_when_value=True),
        ],
    ),
    outputs=[NodeOutput(name="script", type="DICT")],
)
